[VideoFrame] Optimize RGBA -> ARGB conversion.

mstyura opened the pull request:
#### 490d4b4b547992600fdc6a288a8c0d99bd3a8221
<pre>
[VideoFrame] Optimize RGBA -&gt; ARGB conversion.

<a href="https://bugs.webkit.org/show_bug.cgi?id=296145.">https://bugs.webkit.org/show_bug.cgi?id=296145.</a>

Reviewed by Darin Adler.

Employ Accelerate framework to perform RGBA -&gt; ARGB
conversion more efficiently.
The RGBA -&gt; ARGB conversion became around 25 times
faster.

* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::createRGBA):

Canonical link: <a href="https://commits.webkit.org/297806@main">https://commits.webkit.org/297806@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd3829f25e66ce505377b2281164211026074c33

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112838 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32573 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23051 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119041 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63361 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114800 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33225 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41136 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85874 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36526 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115785 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26518 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101507 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66180 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25806 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19639 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62800 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95969 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19713 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122263 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39916 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29760 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94737 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40300 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97727 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94476 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24137 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39603 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17416 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36021 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39802 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45300 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39442 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42775 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41180 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -54,6 +55,92 @@
 #include "CoreVideoSoftLink.h"
 
+
+namespace {
```
WebKit coding style mostly uses `static` for file-private functions rather than an unnamed namespace.

mstyura replied:
Got it, thanks. Changed to static and moved close to other helper function `copyToCVPixelBufferPlane`.

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -54,6 +55,92 @@
+
+namespace {
+
+// Equivalent to libyuv::RGBAToARGB based on Accelerate.framework
```
Not sure why this is an important comment. Is `libyuv` particularly important prior art?

mstyura replied:
I wanted to indicate that the function signature mimics what is done in `libyuv`. I've changed the signature in a way which I hope matches the WebKit approach.

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -54,6 +55,92 @@
+    uint8_t* dst_argb,
+    int dst_stride_argb,
+    int width,
+    int height) {
```
This doesn’t look like WebKit coding style. Naming of arguments is different, format of function definition is also different. Is there a reason this needs different style? Maybe it’s copied from another project? Maybe `libyuv` project style?

Why are the integer types here `int`? The values in the calling function are of type `size_t`, and so this is chopping down a 64-bit unsigned integer into a 32-bit signed integer. And `vImagePixelCount` is a 64-bit unsigned integer, so it’s not even all that helpful to take an integer and then truncate it and then re-expand it. Please instead stick with `size_t` or use `vImagePixelCount` in the function signature.

Why does this function indicate success vs. failure with an int that is `0` for success and `-1` for failure? In WebKit we sometimes use `bool` for this sort of thing, or we could just return the `vImage_Error` directly. I don’t think this style is particularly good.

mstyura replied:
> Maybe libyuv project style?

Yes, I've copy-pasted signatures of `libuv` as was not sure where it goes after code-review, like should I stay on Acccelerate (which is Apple specific) or should I rely on `libuv` to perform such conversion. So I've picked signature which worked well for `libyuv`. Of course it make sense to change it now as long as Accelerate is ok.

mstyura replied:
> format of function definition is also different

I've noticed there is a script to verify the style is correct. Is there something which automatically formats new code according to WebKit style?

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -54,6 +55,92 @@
+    int width,
+    int height) {
+
+    vImage_Buffer src = {
```
WebKit coding style uses words rather than abbreviations for names. So `source` or `sourceBuffer`.

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -54,6 +55,92 @@
+        .rowBytes = static_cast<size_t>(src_stride_rgba)
+    };
+
+    vImage_Buffer dst = {
```
WebKit coding style uses words rather than abbreviations for names. So `destination` or `destinationBuffer`.

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -54,6 +55,92 @@
+        .rowBytes = static_cast<size_t>(dst_stride_argb)
+    };
+
+    // Keep channel order utilizing vImagePermuteChannels_ARGB8888 as optimized plane copy
```
Is `vImagePermuteChannels_ARGB8888` the best optimized plane copy function available in the Accelerate framework?

Did some quick research and I am pretty sure it’s better to use `vImageCopyBuffer`.

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -164,7 +241,12 @@
 
-    copyToCVPixelBufferPlane(rawPixelBuffer, 0, span, height, plane.sourceWidthBytes);
+    std::span<uint8_t> destination = CVPixelBufferGetSpanOfPlane(rawPixelBuffer, 0);
+    uint32_t bytesPerRowDestination = CVPixelBufferGetBytesPerRowOfPlane(rawPixelBuffer, 0);
```
No reason to use uint32_t here when the code above uses size_t for the same thing. We should be consistent.

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -54,6 +55,92 @@
+}
+
+// Equivalent to libyuv::ARGBCopy based on Accelerate.framework
+int ARGBCopy(
```
We don’t need to have an `ARGBCopy` function; it’s not used. Just name this `BGRACopy` instead?

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -164,7 +241,12 @@
+
+    int res = BGRACopy(span.data(), plane.sourceWidthBytes, destination.data(), bytesPerRowDestination, width, height);
+    if (res)
+        return nullptr;
```
Seems to me that if this optimization is good, it should be done inside `copyToCVPixelBufferPlane` not just in `VideoFrame::createBGRA`. I’m pretty sure we can make code that can use `Accelerate` for all cases of `copyToCVPixelBufferPlane`.

mstyura replied:
Yes, seem that it's possible to integrate `Accelerate` directly into `copyToCVPixelBufferPlane` by employing  `vImageCopyBuffer` function which I've initially missed in documentation and used `vImagePermuteChannels_ARGB8888` instead to perform copy BGRA.
But seems like integrating `copyToCVPixelBufferPlane` is more involving having to handle the planar and interleaved buffers, so I'd prefer this change be out of scope of this particular PR and undone anything except RGBA -> BGRA conversion in constructor of `VideoFrame`, so making this PR easier to review and having smaller and clear affected area.

darinadler reviewed (changes requested):
Looks like a good idea; needs at least some changes to fit into WebKit well.

mstyura reviewed (commented):
Thank you very much for review @darinadler. I hope I've fixed most of the problems now.

I've decided to reduce scope of the PR just to optimising the RGBA -> ARGB conversion during video frame construction.
This hopefully make this PR more convenient to review.

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -75,6 +76,30 @@
+        .height = static_cast<vImagePixelCount>(height),
+        .width = static_cast<vImagePixelCount>(width),
+        .rowBytes = bytesPerRowDestination,
+    };
```
For our Safer C++ strategy, I think we need a function that turns the span, bytes per row, width, and height into a vImage_Buffer with range checking that is the smallest wrapper for Accelerate use.

A first step would be to make sure we use a separate function for this, even if for now it’s private here.

```C++
static vImage_Buffer makeVImageBuffer8888(std::span<uint8_t> data, size_t width, size_t height, size_t rowBytes)
{
    constexpr size_t bytesPerPixel = 4;
    RELEASE_ASSERT(width <= rowBytes / bytesPerPixel);
    RELEASE_ASSERT(!height || data.size() / height >= rowBytes);
    return vImage_Buffer {
        .data = data.data(),
        .width = width,
        .height = height,
        .rowBytes = rowBytes,
    };
}
```

I would also ask @cdumez for advice about where he thinks the function should go, but seems fine to land it here at first.

A mildly difficult part is writing the assertions correctly; they are where much of the safety comes from. For example, I wrote them using division rather than multiplication to avoid overflow. I think they are relatively complete.

The only thing about this function that is specific to `8888` is the math in the first assertion, since it’s four bytes per pixel. Otherwise I would have just named it `makeVImageBuffer`.

The key to the safety is the fact that the size of the span is checked by the wrapper so it catches mistakes even if we don’t catch them elsewhere, and also that this carefully written wrapper goes somewhere separate from the rest of the code, separating the simple code that must be checked super carefully from the rest of the code that doesn’t have to be written as carefully and can be checked by the automated Safer C++ checking.

Note that we don’t need `static_cast` to convert `size_t` to `vImagePixelCount` or to convert `uint8_t*` to `uint8_t*`.

Then the code would use it like this:

```suggestion
    auto sourceBuffer = makeVImageBuffer8888(spanConstCast<uint8_t>(sourceRGBA), bytesPerRowSource, width, height);
    auto destinationBuffer = makeVImageBuffer8888(destinationRGBA, bytesPerRowDestination, width, height);
```

mstyura replied:
Thank you very much for the additional round of review! I've incorporated your suggestions, with a few minor modifications that I hope are acceptable.

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -75,6 +76,30 @@
+    vImage_Error err = vImagePermuteChannels_ARGB8888(&sourceBuffer, &destinationBuffer, channelMap, kvImageNoFlags);
+
+    // Permutation does not fail when provided arguments are valid
+    ASSERT_UNUSED(err, err == kvImageNoError);
```
```suggestion
    auto error = vImagePermuteChannels_ARGB8888(&sourceBuffer, &destinationBuffer, channelMap, kvImageNoFlags);

    // Permutation will not fail as long as the provided arguments are valid.
    ASSERT_UNUSED(error, error == kvImageNoError);
```

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -130,19 +155,8 @@
     size_t bytesPerRowDestination = CVPixelBufferGetBytesPerRowOfPlane(rawPixelBuffer, 0);
-    for (unsigned i = 0; i < height; ++i) {
-        size_t j = 0;
-        while (j < std::min(plane.sourceWidthBytes, bytesPerRowDestination)) {
```
This old code ignores `width` entirely when deciding how many bytes to process. Here’s hoping that it’s OK that our newer code does in fact process a number of bytes that’s based on the width.

darinadler reviewed (approved):
I think it’s probably OK to land this as is, but I think it would be better to do the better version of Safer C++ handling I am suggesting here.

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -75,6 +76,38 @@
+    auto error = vImagePermuteChannels_ARGB8888(&sourceBuffer, &destinationBuffer, channelMap, kvImageNoFlags);
+
+    // Permutation will not fail as long as the provided arguments are valid.
+    ASSERT_UNUSED(error, error == kvImageNoError);
```
Not sure that having every other line be blank makes this code easier to read!

```suggestion
    auto sourceBuffer = makeVImageBuffer8888(spanConstCast<uint8_t>(sourceRGBA), width, height, bytesPerRowSource);
    auto destinationBuffer = makeVImageBuffer8888(destinationARGB, width, height, bytesPerRowDestination);
    uint8_t channelMap[4] = { 3, 0, 1, 2 };
    auto error = vImagePermuteChannels_ARGB8888(&sourceBuffer, &destinationBuffer, channelMap, kvImageNoFlags);
    // Permutation will not fail as long as the provided arguments are valid.
    ASSERT_UNUSED(error, error == kvImageNoError);
```

darinadler commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -130,19 +163,8 @@
-        skip(destination, bytesPerRowDestination);
-    }
+
+    copyRGBAToARGB(source, plane.sourceWidthBytes, destination, bytesPerRowDestination, width, height);
```
Not sure that we benefit much from a separate function for this. Consider just putting the 4 lines of code here, and also getting rid of the unneeded local variables `source`, `destination`, and `bytesPerRowDestination`. Also note that we created this pixel buffer and it’s not planar so we can use simpler functions.

```suggestion
    // Convert from RGBA to ARGB.
    auto source = makeVImageBuffer8888(spanConstCast<uint8_t>(span), width, height, plane.sourceWidthBytes);
    auto destination = makeVImageBuffer8888(CVPixelBufferGetSpan(pixelBuffer.get()), width, height, CVPixelBufferGetBytesPerRow(pixelBuffer.get()));
    constexpr uint8_t channelMap[4] = { 3, 0, 1, 2 };
    auto error = vImagePermuteChannels_ARGB8888(&source, &destination, channelMap, kvImageNoFlags);
    ASSERT_UNUSED(error, error == kvImageNoError);
```

Or consider adding a more focused helper function instead:

```C++
static vImage_Buffer makeVImageBuffer8888(CVPixelBufferRef buffer)
{
    return makeVImageBuffer8888(
        CVPixelBufferGetSpan(buffer),
        CVPixelBufferGetWidth(buffer),
        CVPixelBufferGetHeight(buffer),
        CVPixelBufferGetBytesPerRow(buffer)
    );
}
```

```suggestion
    // Convert from RGBA to ARGB.
    auto source = makeVImageBuffer8888(spanConstCast<uint8_t>(span), width, height, plane.sourceWidthBytes);
    auto destination = makeVImageBuffer8888(pixelBuffer.get());
    constexpr uint8_t channelMap[4] = { 3, 0, 1, 2 };
    auto error = vImagePermuteChannels_ARGB8888(&source, &destination, channelMap, kvImageNoFlags);
    ASSERT_UNUSED(error, error == kvImageNoError);
```

darinadler reviewed (approved):
Seems fine to land exactly as is. I have two more comments, both optional.

mstyura committed:
[VideoFrame] Optimize RGBA -> ARGB conversion.

https://bugs.webkit.org/show_bug.cgi?id=296145.

Reviewed by Darin Adler.

Employ Accelerate framework to perform RGBA -> ARGB
conversion more efficiently.
The RGBA -> ARGB conversion became around 25 times
faster.

* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::createRGBA):

Canonical link: https://commits.webkit.org/297806@main
